src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -2728,6 +2745,120 @@ - (BOOL)_attributeAffectsDeviceConfiguration:(MTRAttributePath *)attributePath
+            }
+            break;
+        }
+
+        // If the server list changed and clusters were removed, remove the clusters from the _persistedClusters and _persistedClusterData for that endpoint
+        // Also remove it from the data store.
+        case MTRAttributeIDTypeClusterDescriptorAttributeServerListID: {
+            NSMutableSet * toBeRemovedClusters = [NSMutableSet setWithArray:[self arrayOfNumbersFromAttributeValue:previousValue]];
+            NSSet * clustersOnDevice = [NSSet setWithArray:[self arrayOfNumbersFromAttributeValue:attributeDataValue]];
+            [toBeRemovedClusters minusSet:clustersOnDevice];
+
+            NSMutableSet<MTRClusterPath *> * clusterPathsToRemove = [[NSMutableSet alloc] init];
+            for (NSNumber * cluster in toBeRemovedClusters) {
+                for (MTRClusterPath * path in _persistedClusters) {
+                    if ([path.endpoint isEqualToNumber:attributePath.endpoint] && [path.cluster isEqualToNumber:cluster]) {
+                        [clusterPathsToRemove addObject:path];
+                        [_persistedClusterData removeObjectForKey:path];
+
+                        [self.deviceController.controllerDataStore clearStoredClusterDataForNodeIDWithClusterID:self.nodeID endpointID:path.endpoint clusterID:path.cluster];
+                    }
+                }
+            }
+            [_persistedClusters minusSet:clusterPathsToRemove];
+            break;
+        }
+        }
+    }
+
+    switch (attributePath.attribute.unsignedLongValue) {
+    // If the attribute list changed and attributes were removed, remove the attributes from the _persistedClusterData for that cluster and endpoint.
+    // Also remove it from the data store cluster data.
+    case MTRAttributeIDTypeGlobalAttributeAttributeListID: {
+        NSMutableSet * toBeRemovedAttributes = [NSMutableSet setWithArray:[self arrayOfNumbersFromAttributeValue:[self _cachedAttributeValueForPath:attributePath]]];
+        NSSet * attributesOnDevice = [NSSet setWithArray:[self arrayOfNumbersFromAttributeValue:attributeDataValue]];
+
+        [toBeRemovedAttributes minusSet:attributesOnDevice];
+        for (NSNumber * attribute in toBeRemovedAttributes) {
+            for (MTRClusterPath * path in _persistedClusters) {
+                if ([path.endpoint isEqualToNumber:attributePath.endpoint] && [path.cluster isEqualToNumber:attributePath.cluster]) {
+                    MTRDeviceClusterData * clusterData = [self _clusterDataForPath:path];
+                    if (clusterData == nil) {
+                        return;
+                    }
+                    [clusterData _removeValueForAttribute:attribute];
+                    [self->_persistedClusterData setObject:clusterData forKey:path];
+
+                    NSDictionary<MTRClusterPath *, MTRDeviceClusterData *> * dataStoreClusterData = [self.deviceController.controllerDataStore getStoredClusterDataForNodeID:self.nodeID];
```

Comment:
yeah makes sense but this is not needed anymore if we do the following - 3. Clear the stored value, so that when the _persistedClusterData value is paged in from storage the right thing happens.